### ISSUE TITLE
Premature finalization in gctest (cygwin)

### ISSUE THREAD

**Issue by ivmai** [ivmai/bdwgc#375]
Reproduced on latest master (https://github.com/ivmai/bdwgc/commit/f65701b40cf0ca91b13868940a84a44510e429c4), build https://ci.appveyor.com/project/ivmai/bdwgc/builds/40684362/job/rahbqfcq82hc4lsg
Output (64-bit cygwin):

```
Started a child process, pid=1855
Premature finalization - collector is broken
Test failed
Starting tiny reverse test, pid=1854
Finished a child process, pid=1854
Forked child process, pid=1856
Started a child process, pid=1856
Exception: STATUS_ACCESS_VIOLATION at rip=004C4D4EC66
rax=00000000FFBFC198 rbx=0000000000000000 rcx=0000000001C00002
...
rbp=0000000000000200 rsp=00000000FF3FC950
program=C:\projects\bdwgc\.libs\gctest.exe, pid 1856, thread main
cs=0033 ds=002B es=002B fs=0053 gs=002B ss=002B
Child process failed, pid=1856, status=0x8b
Test failed
Stack trace:
Frame        Function    Args
000FF3FC0B0  00180061979 (00000000000, 00000000740, 7FFC82ED9729, 000FF3FDE50)
...
00000000000  001800D6551 (00000050000, 0000008CC80, 00000003000, 00000000000)
End of stack trace
make: *** [Makefile:2795: check-without-test-driver] Aborted (core dumped)
Command exited with code 2
```

**Comment by ivmai** [comment]
Differs from issues #256

**Comment by ivmai** [comment]
Latest build (x86): https://ci.appveyor.com/project/ivmai/bdwgc/builds/40914357/job/39lts7n3n59c0cmu 
```
./gctest.exe
Forked child process, pid= 2676
Started a child process, pid= 2676
Starting tiny reverse test, pid= 2676
Finished a child process, pid= 2676
Finished reverse_test at time 776 (0xffc5cd70)
Finished typed_test at time 1725 (0xffc5cd70)
Forked child process, pid= 2677
Started a child process, pid= 2677
Forked child process, pid= 2678
Started a child process, pid= 2678
Premature finalization - collector is broken
Test failed
Forked child process, pid= 2679
Started a child process, pid= 2679
Child process failed, pid= 2677, status= 0x86
Test failed
Premature finalization - collector is broken
Test failed
Stack trace:
Frame     Function  Args
End of stack trace
Started a child process, pid= 2680
make: *** [Makefile:2795: check-without-test-driver] Aborted (core dumped)
Premature finalization - collector is broken
Test failed
```

**Comment by ivmai** [comment]
Premature finalization is fixed by commit b8c4e72
